crates/evm/src/query.rs

```diff
@@ -615,37 +513,67 @@ impl<C: sov_modules_api::Context> Evm<C> {
+        let mut block_env = match block_id {
+            Some(BlockId::Number(block_num)) => {
+                match block_num {
+                    BlockNumberOrTag::Pending | BlockNumberOrTag::Latest => {
+                        // if no block is produced yet, should default to genesis block env, else just return the lates
+                        self.block_env.get(working_set).unwrap_or_else(|| {
+                            BlockEnv::from(
+                                &self
+                                    .get_sealed_block_by_number(
+                                        Some(BlockNumberOrTag::Earliest),
+                                        working_set,
+                                    )
+                                    .unwrap()
+                                    .expect("Genesis block must be set"),
                             )
-                            .unwrap()
-                            .expect("Genesis block must be set"),
-                    )
-                })
-            }
-            _ => {
-                let block = match self.get_sealed_block_by_number(block_number, working_set)? {
-                    Some(block) => block,
-                    None => return Err(EthApiError::UnknownBlockNumber.into()),
-                };
+                        })
+                    }
+                    _ => {
+                        let block =
+                            match self.get_sealed_block_by_number(Some(block_num), working_set)? {
+                                Some(block) => block,
+                                None => return Err(EthApiError::UnknownBlockNumber.into()),
+                            };
 
-                set_state_to_end_of_evm_block(block.header.number, working_set);
+                        set_state_to_end_of_evm_block(block.header.number, working_set);
 
-                BlockEnv::from(&block)
+                        BlockEnv::from(&block)
+                    }
+                }
             }
+            Some(BlockId::Hash(block_hash)) => {
+                let block_number = self
+                    .get_block_number_by_block_hash(block_hash.block_hash, working_set)
+                    .ok_or_else(|| EthApiError::UnknownBlockOrTxIndex)?;
+
+                set_state_to_end_of_evm_block(block_number, working_set);
+
+                BlockEnv::from(
+                    &self
+                        .get_sealed_block_by_number(
+                            Some(BlockNumberOrTag::Number(block_number)),
+                            working_set,
+                        )
+                        .unwrap()
+                        .expect("Block must be set"),
+                )
+            }
+            None => self.block_env.get(working_set).unwrap_or_else(|| {
+                BlockEnv::from(
+                    &self
+                        .get_sealed_block_by_number(Some(BlockNumberOrTag::Earliest), working_set)
```

Comment:
```suggestion
                        .get_sealed_block_by_number(Some(BlockNumberOrTag::Latest), working_set)
```
Did you this?

Reply:
If block Id is None we should return latest but in this case since there is no block is produced we return the genesis block env, the same logic in lines 527-535
(If the working set exists it is already in the latest state so we don't need to do anything else)